z3c/dependencychecker/modules.py

```diff
@@ -0,0 +1,398 @@
+# -*- coding: utf-8 -*-
+from xml.etree import ElementTree
+from z3c.dependencychecker.dotted_name import DottedName
+import ast
+import fnmatch
+import os
+import re
+
+
+TEST_REGEX = r"""
+{0}    # Needs to start with the path separator
+\w*    # There could be some characters (or none)
+test   # has to contain test on it
+s?     # ensure plurals are handled as well
+(      # Needs to end with either a path or a file extension:
+{0}    # path separator
+|      # OR
+\w*    # some possible characters
+\.     # a dot
+\w+    # extension
+)
+""".format(os.sep)
+
+TEST_IN_PATH_RE = re.compile(
+    TEST_REGEX,
+    re.VERBOSE,
+)
+
+
+class BaseModule(object):
+
+    def __init__(self, package_path, full_path):
+        self.path = full_path
+        self._relative_path = self._get_relative_path(package_path, full_path)
+        self.testing = self._is_test_module()
+
+    @staticmethod
+    def _get_relative_path(package_path, full_path):
+        return full_path[len(package_path):]
+
+    def _is_test_module(self):
+        return bool(re.search(TEST_IN_PATH_RE, self._relative_path))
+
+    @classmethod
+    def search_files(cls, top_dir):
```

Comment:
See comment on `PythonModule` below regarding name.

Reply:
You'll have to make clearer what the flow is. What does this method do? What does `scan()` do? Which gets called first? There are quite some methods in the class.

I haven't looked at the rest of the code yet, but my *current* guess (and thus also my current naming suggestion) would be:

- `search_files()` is the objects-creating method. Iterate through something and return instantiated `BaseModel` subclasses. An `instantiate_from_dir/package/file/whatever()` might be clearer. Or `create_....()`?

- `scan()` is what then gets called (by what? I would expect it to be called in the `__init__()`?) to read the file and extract info from it. Would `.extract_info()` be a clearer name? Depends also on what it actually does.

Reply:
I'm starting to regret the way I have split the branch... :-/

This commit should shed some light: https://github.com/reinout/z3c.dependencychecker/pull/54/commits/eba1ceb40517e52717876b8f7d641b85e2fcd2c1

In short:

``` 
for module_obj in MODULES:
    for source_file in module_obj.search_files(top_folder):
        self.imports.add_imports(source_file.scan())
```
At the very bottom of modules.py there is a tuple with all the classes defined there. This is looped through (see above) and on each of them the ``search_files`` method is called which returns an iterator with all the files for that given specific extractor that can be found.

For each file found, a new object from the same class is returned and when called on the ``scan`` method it returns an iterator giving you all imports (if any) that have been found on that file.

I thought that a ``@classmethod`` rather than some external method/function would be nicer as it allows us to wrap both functionality (searching specific files and searching imports on those specific files) in a single class. Probably it would make sense to describe this *protocol* on the module global docstring, so that another contributor willing to support whatever files (markdown?) would know what to do.

As for the method names, I'm completely open to change them, but I guess this explanation was needed first.